[SymmetricMemory] support specifying group_name at rendezvous time

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139677
* __->__ #139529

Before this PR, users need to call `empty_strided_p2p()` with a `group_name`:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device=device, group_name="0")
symm_mem = _SymmetricMemory.rendezvous(tensor)
```

Users can now omit `group_name` at allocation time and specify it later at rendezvous time:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device=device)
symm_mem = _SymmetricMemory.rendezvous(tensor, group_name="0")
```

Rationales for this change:
- This allows the same allocation to establish symmetric memory under different groups
- Specifying `group_name` at rendezvous time instead of allocation time is a more natural UX

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Commit:
[SymmetricMemory] support specifying group_name at rendezvous time

[ghstack-poisoned]

Commit:
Update on "[SymmetricMemory] support specifying group_name at rendezvous time"


Before this PR, users need to call `empty_strided_p2p` with a `group_name`:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device="cuda", group_name="0")
symm_mem = _SymmetricMemory.rendezvous(tensor)
```

Users can now omit `group_name` at allocation time and specify it later at rendezvous time:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device="cuda")
symm_mem = _SymmetricMemory.rendezvous(tensor, group_name="0")
```

Rationale for this change:
- This allows the same allocation to participate in multiple symmetric memory formations under different groups
- More natural UX

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

Commit:
Update on "[SymmetricMemory] support specifying group_name at rendezvous time"


Before this PR, users need to call `empty_strided_p2p` with a `group_name`:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device="cuda", group_name="0")
symm_mem = _SymmetricMemory.rendezvous(tensor)
```

Users can now omit `group_name` at allocation time and specify it later at rendezvous time:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device="cuda")
symm_mem = _SymmetricMemory.rendezvous(tensor, group_name="0")
```

Rationale for this change:
- This allows the same allocation to participate in multiple symmetric memory formations under different groups
- More natural UX

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge

Commit:
Update on "[SymmetricMemory] support specifying group_name at rendezvous time"


Before this PR, users need to call `empty_strided_p2p()` with a `group_name`:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device=device, group_name="0")
symm_mem = _SymmetricMemory.rendezvous(tensor)
```

Users can now omit `group_name` at allocation time and specify it later at rendezvous time:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device=device)
symm_mem = _SymmetricMemory.rendezvous(tensor, group_name="0")
```

Rationales for this change:
- This allows the same allocation to establish symmetric memory under different groups
- Specifying `group_name` at rendezvous time instead of allocation time is a more natural UX

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

lw commented on torch/_C/_distributed_c10d.pyi:
```diff
@@ -653,14 +653,17 @@ class _SymmetricMemory:
         dtype: torch.dtype,
         device: torch.device,
-        group_name: str,
+        group_name: str | None = None,
```
Do we really need to keep backwards compatibility? SymmetricMemory is a prototype feature, right?

Then again, I'm probably chill with this idea only because I haven't yet landed my code that makes use of SymmetricMemory :P

yifuwang replied:
> Do we really need to keep backwards compatibility?

No

> SymmetricMemory is a prototype feature, right?

You are right. And this function is a private pybind of a prototype feature. I just wanted make the change less disruptive to early adopters :)

lw commented on torch/csrc/distributed/c10d/CUDASymmetricMemory.cu:
```diff
@@ -592,12 +587,27 @@ int CUDASymmetricMemory::get_world_size() {
 #if !defined(USE_ROCM) && defined(PYTORCH_C10_DRIVER_API_SUPPORTED)
-  auto driver_api = c10::cuda::DriverAPI::get();
+  c10::cuda::CUDAGuard guard(device_idx);
+  device_idx = static_cast<int>(guard.current_device().index());
```
What's the point of this?

yifuwang replied:
Just wanted to move an existing guard up so that future changes relying on cudaSetDevice become more foolproof.

lw reviewed (approved):
I am a tad lost with the indirections between (Cuda)SymmetricMemory, (Cuda)SymmetricMemoryAllocator, AllocationRef and Block. I assume some of it is due to the need to be somewhat device-agnostic in some parts of the code? And, now, to the need for multiple process groups to map to the same tensor?

Apart from that, however, the code seems to make sense to me!

yifuwang commented:
> I am a tad lost with the indirections between (Cuda)SymmetricMemory, (Cuda)SymmetricMemoryAllocator, AllocationRef and Block. I assume some of it is due to the need to be somewhat device-agnostic in some parts of the code?

Yes. `SymmetricMemory` is the device-agnostic, user-facing interface. `SymmetricMemoryAllocator` is the interface for device-specific backends.

`AllocationRef` and `Block` are implementation details of `CUDASymmetricMemory{Allocator}`. `AllocationRef` owns the lifetime of a (`vaddr`, `cumem_handle`) pair. It unmaps the vaddr and releases the cumem_handle upon destruction (`cumem_handle` could be an allocation or an imported handle. It is ref-counted under the hood). `Block` is the metadata for an allocation (mainly for rendezvous).

I'll add some proper documentation to these.

Commit:
Update on "[SymmetricMemory] support specifying group_name at rendezvous time"


Before this PR, users need to call `empty_strided_p2p()` with a `group_name`:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device=device, group_name="0")
symm_mem = _SymmetricMemory.rendezvous(tensor)
```

Users can now omit `group_name` at allocation time and specify it later at rendezvous time:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device=device)
symm_mem = _SymmetricMemory.rendezvous(tensor, group_name="0")
```

Rationales for this change:
- This allows the same allocation to establish symmetric memory under different groups
- Specifying `group_name` at rendezvous time instead of allocation time is a more natural UX

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

Commit:
Update on "[SymmetricMemory] support specifying group_name at rendezvous time"


Before this PR, users need to call `empty_strided_p2p()` with a `group_name`:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device=device, group_name="0")
symm_mem = _SymmetricMemory.rendezvous(tensor)
```

Users can now omit `group_name` at allocation time and specify it later at rendezvous time:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device=device)
symm_mem = _SymmetricMemory.rendezvous(tensor, group_name="0")
```

Rationales for this change:
- This allows the same allocation to establish symmetric memory under different groups
- Specifying `group_name` at rendezvous time instead of allocation time is a more natural UX

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

Commit:
Update on "[SymmetricMemory] support specifying group_name at rendezvous time"


Before this PR, users need to call `empty_strided_p2p()` with a `group_name`:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device=device, group_name="0")
symm_mem = _SymmetricMemory.rendezvous(tensor)
```

Users can now omit `group_name` at allocation time and specify it later at rendezvous time:

```python
tensor = _SymmetricMemory.empty_strided_p2p((1024,), (1,), device=device)
symm_mem = _SymmetricMemory.rendezvous(tensor, group_name="0")
```

Rationales for this change:
- This allows the same allocation to establish symmetric memory under different groups
- Specifying `group_name` at rendezvous time instead of allocation time is a more natural UX

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge
